### ISSUE TITLE
Support Pluto notebooks in Franklin

### ISSUE THREAD

**Issue by tlienart** [tlienart/Franklin.jl#813]
Following a discussion with Fons, here are the short notes:

* effectively https://github.com/mitmath/18S191 already has an integration but
* --> file mgt is a bit awkward
* --> would be nicer to just allow `.jl` files that can be recognised as Pluto files (`is_pluto_notebook`) and if they are, be treated as such (provided Pluto is loaded, we can use Requires for this check)
* (in a similar vein, we should probably use Requires for Literate but Literate is very lightweight)

**Comment by VarLad** [comment]
@tlienart Lets discuss this integration a bit more

**Comment by VarLad** [comment]
My question here is,
Assume we already have converted the notebooks to HTML via github actions, my point is, how do we wrap the header and footer and sidebar around that HTML?

**Comment by tlienart** [comment]
This is not the right path in that there will be a mismatch between local and remote. 

Let's say however that you already have a local workflow that generates HTML files from Pluto notebooks. Let's say that such a file is at path `./pluto/page.html`. Then what you can do is create a secondary file which includes that HTML something like `page.md` with content

```
{{insert ../pluto/page.html}}
```

the `..` is important as `insert` takes paths relative to the `_layout` folder. 

Note: I'm not going to discuss this much more, I know this is a wanted feature and am willing to work on it. It's not entirely trivial and will take me a while to do well.
Feel free to experiment with alternative approaches though but I have limited time to brainstorm here.

**Comment by fonsp** [comment]
@VarLad take a look at the source code for https://github.com/mitmath/18S191

**Comment by StatisticalMice** [comment]
I created an example repo with just basic Franklin and one Pluto notebook, and then converted the notebook to HTML using 

```
julia> using PlutoSliderServer
julia> PlutoSliderServer.export_notebook("penguins-aog-makie.jl")
```

https://github.com/StatisticalMice/franklin-pluto-demo

The 'Penguins' menu item shows the Pluto notebook in HTML, but without Franklin.

The 'Penguins 2' menu item tries to use the `insert` described above, but it doesn't display correctly. I believe the problem is that the HTML is not properly structured, it has two `<head>` sections.

**Comment by tlienart** [comment]
Right, so what you can try for experimentation is to modify the `_layout` files to have an if statement based on whether the current page is a notebook or not. To do that you would add 

```
@def notpluto = false
```

Then in the layout you'd have blocks like

```
{{if notpluto}}
... the usual stuff ...
{{end}}
```

And in configmd you need to add the default `notpluto = true`.

So basically that will do the same as usual in all cases except in the pluto case. Once that works you can start fiddling with what gets included or not to get the appropriate layout

**Comment by atiyabzafar** [comment]
I tried experimentation as mentioned by you @tlienart for adding pluto notebook. But it was turning out to be a hassle. In head.html I used 
```
{{if notpluto}}
... the usual stuff ...
{{end}}
```
I have earlier used this method to add cytoscape.js functionality in a site. But I simply was not able to get it. Easier way to do that is just to add the html in the github repo and just point a link to that. But that is a cheap way to do this. 

Best and quickest way right now that I implemented was using [PlutoStaticHTML.jl](https://github.com/rikhuijzer/PlutoStaticHTML.jl) and add mathjax using `@def hasmathjax=true` instead of KaTeX. 
This works for now but If i get time in the coming days I will try your way. As you mentioned above it is not a trivial thing to do. Adding this comment for any future visitor as a method to add pluto notebook in Franklin site.

**Comment by vchuravy** [comment]
So I had a quick go at this for my personal website and while not perfect yet it is quite nice.

Using https://github.com/JuliaPluto/PlutoSliderServer.jl/pull/127 + a github action like:

https://github.com/vchuravy/vchuravy.github.io/blob/c6e13351c86f5d1717d3c864d7cda92b022d9186/.github/workflows/Deploy.yml#L33

I generate from:

```
talks/notebook.jl
```

```
talks/notebook/index.html
talks/notebook/notebook.jl
```

and then franklin just copies those files over. I then use the generated json to generate a talks page https://github.com/vchuravy/vchuravy.github.io/blob/c6e13351c86f5d1717d3c864d7cda92b022d9186/utils.jl#L143

x-ref: https://github.com/tlienart/Franklin.jl/issues/813

What I would love be able to do is:
- [ ] Better frontmatter support in Pluto, so that the title etc is not dependent on the filename
- [ ] Styling in the generated html: E.g. I want my top navigation page + some json injection

**Comment by roflmaostc** [comment]
@vchuravy have you done any code to include your top navigation page?
Yours looks almost like what I want :)

**Comment by vchuravy** [comment]
You mean:

https://vchuravy.dev/talks/?

That's

```

"""
    {{talks}}

Plug in the list of talks contained in the `/talks/` folder.
"""
function hfun_talks()
    pluto_json = "talks/pluto_export.json"
    if !isfile(pluto_json)
        return ""
    end
    info = JSON.Parser.parsefile("talks/pluto_export.json")
    notebooks = info["notebooks"]

    io = IOBuffer()
    for (name, data) in notebooks
       title = data["frontmatter"]["title"]
       clean_title = replace(title, "_" => " ")
       write(io, "- [$clean_title](/talks/$title)")
    end
    r = Franklin.fd2html(String(take!(io)), internal=true)
    return r
end
```

I have been meaning to improve that by using frontmatter a bit better and as an example sort by dates and maybe even dates.

**Comment by roflmaostc** [comment]
Yes I meant that!
But unfortunately the notebooks then "leave" your website and the header is gone, right?

**Comment by fonsp** [comment]
> But unfortunately the notebooks then "leave" your website and the header is gone, right?

Pluto has the ability to be **embedded inside webpages**, either with or without iframe. 

## Options
With iframe might be the easiest to implement without going into details of how Pluto works. You can ask github.com/adrhill how to do this :)


Without iframe looks and performs the best, this is what I use in https://plutojl.org/en/docs/javascript-api/ (to get a sidebar on the left). (This site was generated with PlutoPages.jl, but I don't think PlutoPages.jl is a good option for use inside Franklin but feel free to try.) For this you need to know a bit how Pluto HTML files work.

## How Pluto HTML files work
Basically, Pluto or PlutoSliderServer generates export HTML that looks like this: (go to a notebook export like [this one](https://featured.plutojl.org/basic/turtles-art.html), then Right click -> View page source)

```html
<!DOCTYPE html>
<html lang="en">

<head>
    <meta name="viewport" content="width=device-width" />
    <meta charset="utf-8" />

    <!-- ....All frontmatter meta tags... -->

<meta name="pluto-insertion-spot-meta">


    <!-- All JS and CSS imports -->

    <script data-pluto-file="launch-parameters">
window.pluto_notebook_id = undefined;
window.pluto_isolated_cell_ids = undefined;
window.pluto_notebookfile = "turtles-art.jl";
window.pluto_disable_ui = true;
window.pluto_slider_server_url = "https://fonsi.armada.silentech.gr";
window.pluto_binder_url = "https://mybinder.org/v2/gh/fonsp/pluto-on-binder/v0.20.4";
window.pluto_statefile = "turtles-art.plutostate";
window.pluto_preamble_html = undefined;
</script>

<meta name="pluto-insertion-spot-parameters">

    <!-- More JS and CSS imports -->

</head>

<body class="loading no-MαθJax">
    <div style="display: flex; min-height: 100vh;">
        <pluto-editor class="fullscreen">
            <progress style="filter: grayscale(1)" class="statefile-fetch-progress delete-me-when-live" max="100"></progress>
        </pluto-editor>
    </div>
</body>

</html>
```

Pluto notebooks load as follows:

Once `editor.js` gets loaded, it searches for all `pluto-editor` elements on the page, and replaces them with live notebook based on their "launch parameters", which tell the component which statefile to load, etc.

Launch parameters can be set globally (eg `window.pluto_statefile`) or (more strongly) as component attributes (eg `<pluto-editor statefile='../something.plutostate' >`), as done in the PlutoPages output like [this one](https://plutojl.org/en/docs/javascript-api/).

## Header, sidebar
You can just add additional content to the HTML document, and it will not be overwritten by Pluto. Wherever you place the `pluto-editor` component is where the notebook will be rendered. Our CSS is designed to make the `pluto-editor` component responsive: it will resize to fit. Again, check the HTML files generated by PlutoPages to see this in action.

So a good option might be to use Pluto/PlutoSliderServer to generated the HTML files, and then use Gumbo/Regex to put this Pluto content together with your other content. Just be sure that the `<head>` content from Pluto is in the final `<head>`, and that there is a `<pluto-editor>` component somewhere in the`<body>`.
